src/backend/distributed/planner/insert_select_planner.c

```diff
@@ -760,22 +739,22 @@ DistributedInsertSelectSupported(Query *queryTree, RangeTblEntry *insertRte,
 			{
 				return error;
 			}
-
-			/*
-			 * We expect partition column values come from colocated tables. Note that we
-			 * skip this check from the reference table case given that all reference tables
-			 * are already (and by default) co-located.
-			 */
-			if (!TablesColocated(insertRte->relid, selectPartitionColumnTableId))
-			{
-				return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
-									 "INSERT target table and the source relation of the SELECT partition "
-									 "column value must be colocated in distributed INSERT ... SELECT",
-									 NULL, NULL);
-			}
 		}
 	}
 
+	/* All tables in source list and target table should be colocated. */
+	List *distributedRelationIdList = DistributedRelationIdList(subquery);
+	distributedRelationIdList = lappend_oid(distributedRelationIdList,
+											targetRelationId);
+
+	if (!AllDistributedRelationsInListColocated(distributedRelationIdList))
+	{
+		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+							 "INSERT target relation and all source relations of the "
+							 "SELECT must be colocated in distributed INSERT ... SELECT",
+							 NULL, NULL);
+	}
```

Comment:
I'd place this block right after the following because we distributed tables cannot appear in the query otherwise?
```c
		if (subqueryRteListProperties->hasCitusLocalTable ||
			subqueryRteListProperties->hasMaterializedView)
```